### ISSUE TITLE
Package the code we have so far.

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/multidms#12]
# Packaging `multidms`
Here, I'll outline the interface, architecture and packaging todo's for `multidms`

## TODO

- [x] Issue Draft
- [x] Confirm Architecture w/ Erick + Hugh
- [x] Confirm Architecture w/ Jesse
- [x] Stubbing
- [x] Confirm Stubs w/ Erick + Hugh
- [x] Implementation!
- [x] Testing: pytest? 
- [x] pyproject.toml
- [ ] Dockerfile
- [x] Licence # MIT? 
- [x] .gitignore update
- [x] .pre-commit-config.yaml # formatting = Black + flake8
- [x] README # Short description, installation, and usage
- [x] Documentation
- [x] CI/CD # github actions
- [x] release 0.0.1 on pip

## Interface

In designing the API's general interface, we will mimic the one from `dms-variants`, except we will handle the `binarymap` internally, along with a new step that converts all mutations to be relative to the wildtype sequence of the reference condition.

Running `multidms` will involve the following basic steps. First, we import `multidms`:

```python
>>> import pandas as pd
>>> import multidms
```

Next, we read in a dataframe of functional scores for all variants from all conditions to be compared, where rows are variants and columns give the following variant-specific data:
1. amino-acid substitutions (where all homologs/conditions use the same site-numbering scheme)
2. functional score
3. condition (e.g., values might be "Delta" or "Omicron"). 

Note: There can be any number of conditions, each non-reference condition will be given it's own set of non-zero shift parameters in the model.

```python
>>> func_score_df = pd.read_csv('combined_experiment_variant_scores.csv')
```

Next, we use the above dataframe to initialize an object corresponding to a specific model type. For instance, below, we initialize a sigmoid-based model, though we could also include other model types (e.g., linear, spline). This step converts all mutations to be relative to the wildtype sequence of the reference condition, and then generates a binary map, storing it for downstream fitting. It also initializes model parameters, which are retrievable from the object at any time before or after fitting.

```python
>>> multimodel = multidms.globalepistasis.SigmoidEpistasis(
...     func_score_df,
...     condition_col="homolog",
...     reference_condition="Delta",
...     substitution_column="ref_aa_substitutions",
... )
```

The above object will have a `fit()` method, which we will call to fit the model to the data:

```
>>> fit_df = multimodel.fit(target="functional_score")
```

This method will return a dataframe with some useful stats from `jax.opt` (e.g., fitting time, state, loss over time). The model object will also store the results of fitting, including: the predicted functional scores, beta and shift parameters, and all other parameters used to define the model.

The model object will also have methods that plot data, such as a scatter plot comparing observed and predicted functional scores, or a histogram of the distribution of beta and shift parameters, to give two examples.

## Class structure

**Sub Modules**: Here is the high-level outline for the modules that I imagine will be included in this package.

`multidms.globalepistasis`
1. `class AbstractEpistasis(abc.ABC)`
2. `class SigmoidEpistasis(AbstractEpistasis)` # a child class for fitting data with a sigmoidal global-epistasis function
3. `class NoEpistasis(AbstractEpistasis)` # a child class for fitting data with a linear function

As in `dms_variants`, the above classes will have methods that allow the user to plot data in various ways. So far, we are planning to include the following methods:
1. `def func_score_pred_true_scatter` # a scatter plot of observed vs. predicted functional scores, and observed functional score vs. predicted latent phenotype.
2. `def param_histogram` # a histogram of beta and shift parameters
3. `def param_heatmap` # heat maps of beta and shift parameters
6. `def param_agg_by_site` # a line plot showing the sum of the absolute value of shift parameters by site

`multidms.utils`
1. `def split_sub()` # a function for parsing the wildtype amino acid, site number, and mutant amino acid from a string.
2. `def split_subs()` # a function for parsing individual substitutions from a string specifying all substitutions in a variant.

`multidms.constants`
Not sure what we need here yet, but we'll certainly have constants that will need to be defined.

`multidms.error` _Would be nice_ 
1. def multidms_fit_error(Exception):
2. def multidms_substitution_error(Exception):

`multidms.logging` _Would be nice_
Write a logging function that supports verbosity.

## Misc.
* plotting code should be done with altair to make interactive plots
* we want to replicate jesse's [regex](https://github.com/jbloomlab/binarymap/blob/c255e809640298bc7a19e78d9f0e3fc9634a38a5/binarymap/binarymap.py#L466)
* get_mutational_effects() returns a dataframe with a row for each mutation including
  * which conditions it appeas in
  * beta
  * functional effect
  * shift (for each condition)
